### PR TITLE
Add user authentication

### PR DESCRIPTION
**Pull request by archferns**
Adds a User model 
Add Authorisation via HTTP Token auth to the controller.

The expected format for this is as follows:
```
  Authorization: Bearer SECRETTOKEN, user_id=USER-UUID
```
This allows us to fetch the user by the id and then verify by the token
and avoids the need for storing the `SECRETTOKEN`
in plaintext or encrypted, since it is stored as a BCrypt digest
using the standard rails `has_secure_password` method.

### CODE REVIEW

File: spec/controllers/application_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe ApplicationController, type: :controller do
```
Comment (scruti) [inline spec/controllers/application_controller_spec.rb]: Very nicely tested 👏

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -10,4 +13,15 @@ class ApplicationController < ActionController::API
       render json: { error: error }, status: status
     end
   end
+
+  private
+
+  def authenticate
+    authenticate_with_http_token do |token, options|
+      user = User.find_by(id: options[:user_id])
+      @current_user = user&.authenticate_auth_token(token)
```
Comment (scruti) [inline app/controllers/application_controller.rb]: Small question about the variable scope:
does `@current_user` needs to be an instance var in our service? 
I know it tends to be the default but I suspect we only need it  in `authenticate` method scope and don't use it anywhere else...

Comment (archferns) [reply app/controllers/application_controller.rb]: Yep, that is very good point about it not being needed outside the scope of this check. Ill refactor it.

File: spec/controllers/application_controller_spec.rb (hunk 1 above)
Comment (archferns) [reply spec/controllers/application_controller_spec.rb]: ty :bowing_man:

**Pull request merged**